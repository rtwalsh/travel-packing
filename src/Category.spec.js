import { tick } from 'svelte';
import { cleanup, fireEvent, render, wait } from '@testing-library/svelte';

import Category from './Category.svelte';

describe('Category', () => {
    let itemCount = 0;

    const category = { id: 1, name: 'Clothes', items: {} };
    const categories = [ category ];
    const dnd = {};
    const props = { categories, category, dnd, show: 'all' };

    beforeEach(() => {
        category.items = {
            1: { id: 1, name: 'socks', packed: true },
            2: { id: 2, name: 'shoes', packed: false }
        };
        itemCount = Object.keys(category.items).length;
    });

    afterEach(cleanup);

    test('should match snapshot', () => {
        const { container } = render(Category, props);
        expect(container).toMatchSnapshot();
    });

    function expectItemCount(count) {
        return wait(() => {
            const lis = document.querySelectorAll('li');
            expect(lis.length).toBe(count);
        });
    }

    test('should render', async () => {
        const { getByText } = render(Category, props);
        expect(getByText('Clothes'));
        expect(getByText('1 of 2 remaining'));
        expect(getByText('New Item'));
        expect(getByText('Add Item'));
        await expectItemCount(itemCount);
    });

    test('should add an item', async () => {
        const { getByTestId, getByText } = render(Category, props);

        const input = getByTestId('item-input');
        const value = 't-shirts';
        fireEvent.input(input, { target: {value} });
        fireEvent.click(getByText('Add Item'));
        
        await expectItemCount(itemCount + 1);
        expect(getByText(value));
    });

    test('should delete an item', async () => {
        const { getAllByTestId } = render(Category, props);

        const deleteBtns = getAllByTestId('delete');
        fireEvent.click(deleteBtns[0]); // deletes the first item
        await expectItemCount(itemCount - 1);
    });

    test('should toggle an item', async () => {
        const { container, getByText } = render(Category, props);
        
        const checkboxes = container.querySelectorAll('input[type="checkbox"]');
        expect(checkboxes.length).toBe(2);

        const [shoesCheckbox, socksCheckbox ] = checkboxes;

        expect(socksCheckbox.nextElementSibling.textContent).toBe('socks');
        fireEvent.click(socksCheckbox);
        await tick();
        expect(getByText('2 of 2 remaining'));

        expect(shoesCheckbox.nextElementSibling.textContent).toBe('shoes');
        fireEvent.click(shoesCheckbox);
        await tick();
        expect(getByText('1 of 2 remaining'));
    });
});
