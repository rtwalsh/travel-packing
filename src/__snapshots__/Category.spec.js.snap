// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category should match snapshot 1`] = `
<body>
  <div>
    <section
      class="svelte-1fx2ghy"
      ondragover="return false"
    >
      <h3
        class="svelte-1fx2ghy"
      >
        <span>
          Clothes
        </span>
         
        <span
          class="status svelte-1fx2ghy"
        >
          1 of 2 remaining
        </span>
         
        <button
          class="icon svelte-1fx2ghy"
        >
          ðŸ—‘
        </button>
      </h3>
       
      <form>
        <label>
          New Item
            
          <input
            class="svelte-1fx2ghy"
            data-testid="item-input"
            required=""
          />
        </label>
         
        <button
          class="svelte-1fx2ghy"
        >
          Add Item
        </button>
      </form>
       
      <ul
        class="svelte-1fx2ghy"
      >
        <div>
          <li
            class="svelte-4o8dc6"
          >
            <input
              class="svelte-4o8dc6"
              type="checkbox"
            />
             
            <span
              class="packed-false svelte-4o8dc6"
              draggable="true"
            >
              shoes
            </span>
             
            <button
              class="icon svelte-4o8dc6"
              data-testid="delete"
            >
              ðŸ—‘
            </button>
          </li>
           
        </div>
        <div>
          <li
            class="svelte-4o8dc6"
          >
            <input
              class="svelte-4o8dc6"
              type="checkbox"
            />
             
            <span
              class="packed-true svelte-4o8dc6"
              draggable="true"
            >
              socks
            </span>
             
            <button
              class="icon svelte-4o8dc6"
              data-testid="delete"
            >
              ðŸ—‘
            </button>
          </li>
           
        </div>
      </ul>
    </section>
     
    <dialog
      class="dialog svelte-1safvr0"
    >
      <header
        class="svelte-1safvr0"
      >
         
        <div
          class="title svelte-1safvr0"
        >
          Category
        </div>
         
        <button
          class="close-btn svelte-1safvr0"
        >
          âœ–
        </button>
      </header>
       
      <main
        class="svelte-1safvr0"
      >
        <div>
          
        </div>
      </main>
    </dialog>
  </div>
</body>
`;
